src/OrleansRuntime/Messaging/GatewayAcceptor.cs

```diff
@@ -41,7 +41,10 @@ protected override bool RecordOpenedSocket(Socket sock)
             {
                 //convert handshake cliendId to a GeoClient ID 
                 if (!string.IsNullOrEmpty(Silo.CurrentSilo.ClusterId))
+                {
                     client = GrainId.NewClientId(client.PrimaryKey, Silo.CurrentSilo.ClusterId);
+                    adjustReturnAddressOnNextMessage = true;
```

Comment:
Right now it is not possible because the client does not send a second message before the first reply has come back. This could of course change once people modify the client (e.g. to optimize connection initialization by parallelizing these requests).

Anyway, given the other bug you spotted, I have now changed the mechanism so it always translates the address if needed, not just on the first message, so this particular risk should be eliminated.